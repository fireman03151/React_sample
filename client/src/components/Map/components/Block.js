import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { createSelector } from 'reselect';
import { Link } from 'gatsby';

import ga from '../../../analytics';
import { makeExpandedBlockSelector, toggleBlock } from '../redux';
import { userSelector } from '../../../redux';
import Caret from '../../icons/Caret';
import { blockNameify } from '../../../../utils/blockNameify';
/* eslint-disable max-len */
import GreenPass from '../../../templates/Challenges/components/icons/GreenPass';
import GreenNotCompleted from '../../../templates/Challenges/components/icons/GreenNotCompleted';
/* eslint-enable max-len */
const mapStateToProps = (state, ownProps) => {
  const expandedSelector = makeExpandedBlockSelector(ownProps.blockDashedName);

  return createSelector(
    expandedSelector,
    userSelector,
    (isExpanded, { completedChallenges = [] }) => ({
      isExpanded,
      completedChallenges: completedChallenges.map(({ id }) => id)
    })
  )(state);
};

const mapDispatchToProps = dispatch =>
  bindActionCreators({ toggleBlock }, dispatch);

const propTypes = {
  blockDashedName: PropTypes.string,
  challenges: PropTypes.array,
  completedChallenges: PropTypes.arrayOf(PropTypes.string),
  intro: PropTypes.shape({
    fields: PropTypes.shape({ slug: PropTypes.string.isRequired }),
    frontmatter: PropTypes.shape({
      title: PropTypes.string.isRequired,
      block: PropTypes.string.isRequired
    })
  }),
  isExpanded: PropTypes.bool,
  toggleBlock: PropTypes.func.isRequired
};

const mapIconStyle = { height: '15px', marginRight: '10px', width: '15px' };

export class Block extends Component {
  constructor(...props) {
    super(...props);

    this.handleBlockClick = this.handleBlockClick.bind(this);
    this.handleChallengeClick = this.handleChallengeClick.bind(this);
    this.renderChallenges = this.renderChallenges.bind(this);
  }

  handleBlockClick() {
    const { blockDashedName, toggleBlock } = this.props;
    ga.event({
      category: 'Map Block Click',
      action: blockDashedName
    });
    return toggleBlock(blockDashedName);
  }

  handleChallengeClick(slug) {
    return () => {
      return ga.event({
        category: 'Map Challenge Click',
        action: slug
      });
    };
  }

  renderCheckMark(isCompleted) {
    return isCompleted ? (
      <GreenPass style={mapIconStyle} />
    ) : (
      <GreenNotCompleted style={mapIconStyle} />
    );
  }

  renderChallenges(intro = {}, challenges = []) {
    // TODO: Split this into a Challenge Component and add tests
    // TODO: The styles badge and map-badge on the completion span do not exist
    return [intro].concat(challenges).map((challenge, i) => {
      const completedClass = challenge.isCompleted
        ? ' map-challenge-title-completed'
        : '';
      return (
        <li
          className={'map-challenge-title' + completedClass}
          key={'map-challenge' + challenge.fields.slug}
        >
          <span className='badge map-badge'>
            {i !== 0 && this.renderCheckMark(challenge.isCompleted)}
          </span>
          <Link
            onClick={this.handleChallengeClick(challenge.fields.slug)}
            to={challenge.fields.slug}
          >
            {challenge.title || challenge.frontmatter.title}
          </Link>
        </li>
      );
    });
  }

  render() {
    const {
      blockDashedName,
      completedChallenges,
      challenges,
      isExpanded,
      intro
    } = this.props;
    let completedCount = 0;
    const challengesWithCompleted = challenges.map(challenge => {
      const { id } = challenge;
      const isCompleted = completedChallenges.some(
        completedId => id === completedId
      );
      if (isCompleted) {
        completedCount++;
      }
      return { ...challenge, isCompleted };
    });
    return (
      <li className={`block ${isExpanded ? 'open' : ''}`}>
        <button
          aria-expanded={isExpanded}
          className='map-title'
          onClick={this.handleBlockClick}
        >
          <Caret />
          <h4>{blockNameify(blockDashedName)}</h4>
          <div className='map-title-completed'>
            <span>
              {this.renderCheckMark(
                completedCount === challengesWithCompleted.length
              )}
            </span>
            <span>{`${completedCount}/${challengesWithCompleted.length}`}</span>
          </div>
        </button>
        <ul>
          {isExpanded
            ? this.renderChallenges(intro, challengesWithCompleted)
            : null}
        </ul>
      </li>
    );
  }
}

Block.displayName = 'Block';
Block.propTypes = propTypes;

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Block);
