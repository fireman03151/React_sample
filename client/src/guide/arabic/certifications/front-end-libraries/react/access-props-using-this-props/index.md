---
title: Access Props Using this.props
localeTitle: وصول الدعائم باستخدام هذا. المحاصيل
---
## \## Access Props باستخدام هذا .props

للحصول على هذا الحل ، يجب أن تتذكر أولاً كيفية إضافة الأدوات الدعائية إلى المكون الرئيسي:

\`\` \`JSX

 `Once you have the prop set, you can use this.props in your child component. 
` 

أتش تي أم أل

كلمة المرور المؤقتة هي: **{this.props.tempPassword}**

\`\` \`