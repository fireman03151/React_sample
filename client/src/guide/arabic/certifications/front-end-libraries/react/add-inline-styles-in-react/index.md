---
title: Add Inline Styles in React
localeTitle: إضافة أنماط مضمنة في React
---
## إضافة أنماط مضمنة في React

يمكنك تعريف نمط مكون تمرير الكائن مباشرة كـ "نمط" prop. فقط تذكر أن كل خاصية من كائن النمط هي camelcased. لذا ، تُعلن الخصائص مثل "font-size" بأنها "fontSize" لتكون خاصية كائن جافا سكريبت صالحة.

### المفسد

 `class Colorful extends React.Component { 
  render() { 
    // change code below this line 
    return ( 
      <div style={{color: "yellow", fontSize: 24}}>Style Me!</div> 
    ); 
    // change code above this line 
  } 
 }; 
` 

### مصادر

[DOM Elements Style](https://reactjs.org/docs/dom-elements.html#style)