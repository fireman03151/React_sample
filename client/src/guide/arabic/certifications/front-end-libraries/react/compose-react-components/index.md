---
title: Compose React Components
localeTitle: تأليف React Components
---
## تأليف React Components

### ملحوظة

استخدام المكونات المتداخلة كما في challemge السابقة لتقديم المكونات.

### حل

ما يلي هو الحل لل chakkenge ، حيث تجعل الحمضيات و NonCitrus في مكون ثم يتم تقديمها في آخر:

 `class Fruits extends React.Component { 
  constructor(props) { 
    super(props); 
  } 
  render() { 
    return ( 
      <div> 
        <h2>Fruits:</h2> 
        <NonCitrus /> 
        <Citrus /> 
      </div> 
    ); 
  } 
 }; 
 
 class TypesOfFood extends React.Component { 
  constructor(props) { 
     super(props); 
  } 
  render() { 
    return ( 
      <div> 
        <Fruits /> 
        <Vegetables /> 
      </div> 
    ); 
  } 
 }; 
` 

### روابط ذات صلة:

*   [المكونات والدعائم](https://reactjs.org/docs/components-and-props.html)
*   [المكونات المتداخلة](http://www.reactjstutorial.net/nested-components.html)