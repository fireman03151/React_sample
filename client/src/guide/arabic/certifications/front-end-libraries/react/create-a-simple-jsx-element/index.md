---
title: Create a Simple JSX Element
localeTitle: إنشاء عنصر بسيط JSX
---
## إنشاء عنصر بسيط JSX

هذا التحدي بسيط إلى حد ما ، واستبدل div في عنصر JSX مع h1 `javascript const JSX=<h1></h1>;` ثم أضف النص المطلوب. `javascript const JSX = <h1>Hello JSX!</h1>;`