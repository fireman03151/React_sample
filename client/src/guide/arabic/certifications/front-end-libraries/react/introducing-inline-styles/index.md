---
title: Introducing Inline Styles
localeTitle: إدخال الأنماط الداخلية
---
## إدخال الأنماط الداخلية

## حل

يمكن أن يكون هذا الأمر صعبًا قليلاً لأن JSX يشبه إلى حد بعيد HTML ولكن **ليس نفسه** .

لنستعرض الخطوات حتى تعرف الفرق. قم أولاً بتعيين علامة النمط على **كائن JavaScript** .

 `class Colorful extends React.Component { 
  render() { 
    return ( 
      <div style={{}}> 
        Big Red 
      </div> 
    ); 
  } 
 }; 
` 

الآن لديك تعيين علامة النمط الخاص بك إلى كائن فارغ. لاحظ كيف أن هناك مجموعتين من الأقواس المتعرجة. هذا اختلاف مهم بين JSX و HTML.

ثانيًا ، لنقم بتعيين اللون إلى اللون الأحمر.

 `class Colorful extends React.Component { 
  render() { 
    return ( 
      <div style={{ color: 'red' }}> 
        Big Red 
      </div> 
    ); 
  } 
 }; 
` 

أخيرًا ، لنقم بتعيين حجم الخط إلى 72 بكسل.

### المفسد

 `class Colorful extends React.Component { 
  render() { 
    return ( 
      <div style={{ color: 'red', fontSize: '72'}}> 
        Big Red 
      </div> 
    ); 
  } 
 }; 
` 

لاحظ كيف أن سمة **JSX** هي **camelCase** . هذا اختلاف مهم آخر يجب تذكره حول JSX. بالإضافة إلى ذلك ، ربما لاحظت عدم وجود وحدة. في JSX ، عند تعيين سمة fontSize ، تكون **الوحدة اختيارية** وسيتم تعيينها تلقائيًا على px إذا لم يتم تعيينها يدويًا.