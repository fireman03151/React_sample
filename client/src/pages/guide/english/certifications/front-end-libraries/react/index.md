---
title: React
---
## React

![React](https://upload.wikimedia.org/wikipedia/commons/thumb/a/a7/React-icon.svg/200px-React-icon.svg.png)

React or ReactJS is a front-end library which is used to create beautiful, reusable and interactive User Interfaces.

React is written in JavaScript and JSX. It is designed with reusability in mind and follows a component based pattern, which means write once and use infinitely.

It is available as a CDN for browser based development or as a npm package for Command line shell based editing.

React has been developed under the Facebook Open Source project, and is actively maintained by the Open Source community.

#### More Information:
<!-- Please add any articles you think might be helpful to read before writing the article -->
1. [React](https://reactjs.org/)
2. [React Open Source page](https://github.com/facebook/react/)
3. [React Forum](https://discuss.reactjs.org/)
