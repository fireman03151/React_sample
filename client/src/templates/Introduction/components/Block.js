import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { createSelector } from 'reselect';
import { withTranslation } from 'react-i18next';

import { makeExpandedBlockSelector, toggleBlock } from '../redux';
import { completedChallengesSelector, executeGA } from '../../../redux';
import Challenges from './Challenges';
import Caret from '../../../assets/icons/Caret';
import GreenPass from '../../../assets/icons/GreenPass';
import GreenNotCompleted from '../../../assets/icons/GreenNotCompleted';
import { Spacer } from '../../../components/helpers';

const mapStateToProps = (state, ownProps) => {
  const expandedSelector = makeExpandedBlockSelector(ownProps.blockDashedName);

  return createSelector(
    expandedSelector,
    completedChallengesSelector,
    (isExpanded, completedChallenges) => ({
      isExpanded,
      completedChallenges: completedChallenges.map(({ id }) => id)
    })
  )(state);
};

const mapDispatchToProps = dispatch =>
  bindActionCreators({ toggleBlock, executeGA }, dispatch);

const propTypes = {
  blockDashedName: PropTypes.string,
  challenges: PropTypes.array,
  completedChallenges: PropTypes.arrayOf(PropTypes.string),
  executeGA: PropTypes.func,
  isExpanded: PropTypes.bool,
  superBlockDashedName: PropTypes.string,
  t: PropTypes.func,
  toggleBlock: PropTypes.func.isRequired
};

const mapIconStyle = { height: '15px', marginRight: '10px', width: '15px' };

export class Block extends Component {
  constructor(...props) {
    super(...props);

    this.handleBlockClick = this.handleBlockClick.bind(this);
  }

  handleBlockClick() {
    const { blockDashedName, toggleBlock, executeGA } = this.props;
    executeGA({
      type: 'event',
      data: {
        category: 'Map Block Click',
        action: blockDashedName
      }
    });
    return toggleBlock(blockDashedName);
  }

  renderCheckMark(isCompleted) {
    return isCompleted ? (
      <GreenPass style={mapIconStyle} />
    ) : (
      <GreenNotCompleted style={mapIconStyle} />
    );
  }

  renderBlockIntros(arr) {
    return (
      <div className='block-description'>
        {arr.map((str, i) => (
          <p dangerouslySetInnerHTML={{ __html: str }} key={i} />
        ))}
      </div>
    );
  }

  render() {
    const {
      blockDashedName,
      completedChallenges,
      challenges,
      isExpanded,
      superBlockDashedName,
      t
    } = this.props;

    let completedCount = 0;
    const challengesWithCompleted = challenges.map(challenge => {
      const { id } = challenge;
      const isCompleted = completedChallenges.some(
        completedId => id === completedId
      );
      if (isCompleted) {
        completedCount++;
      }
      return { ...challenge, isCompleted };
    });

    const isProjectBlock = challenges.some(challenge => {
      const isJsProject =
        challenge.order === 10 && challenge.challengeType === 5;

      const isOtherProject =
        challenge.challengeType === 3 ||
        challenge.challengeType === 4 ||
        challenge.challengeType === 10;

      const isTakeHomeProject = blockDashedName === 'take-home-projects';

      return (
        (isJsProject && !isTakeHomeProject) ||
        (isOtherProject && !isTakeHomeProject)
      );
    });

    const blockIntroObj = t(
      `intro:${superBlockDashedName}.blocks.${blockDashedName}`
    );
    const blockTitle = blockIntroObj ? blockIntroObj.title : null;
    const blockIntroArr = blockIntroObj ? blockIntroObj.intro : [];

    return isProjectBlock ? (
      <div className='block'>
        <h3>{blockTitle}</h3>
        {this.renderBlockIntros(blockIntroArr)}
        <Challenges
          challengesWithCompleted={challengesWithCompleted}
          isProjectBlock={isProjectBlock}
        />
      </div>
    ) : (
      <div className={`block ${isExpanded ? 'open' : ''}`}>
        <h3>{blockTitle}</h3>
        {this.renderBlockIntros(blockIntroArr)}
        <button
          aria-expanded={isExpanded}
          className='map-title'
          onClick={this.handleBlockClick}
        >
          <Caret />
          <h4>
            {t(`intro:misc-text.section-courses`, { section: blockTitle })}
          </h4>
          <div className='map-title-completed'>
            {this.renderCheckMark(
              completedCount === challengesWithCompleted.length
            )}
            <span className='map-completed-count'>{`${completedCount}/${challengesWithCompleted.length}`}</span>
          </div>
        </button>
        {isExpanded && (
          <Challenges
            challengesWithCompleted={challengesWithCompleted}
            isProjectBlock={isProjectBlock}
          />
        )}
        <Spacer size={2} />
      </div>
    );
  }
}

Block.displayName = 'Block';
Block.propTypes = propTypes;

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withTranslation()(Block));
