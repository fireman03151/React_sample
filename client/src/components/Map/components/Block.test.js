/* global jest, expect */

import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { Block } from './Block';
import mockChallengeNodes from '../../../__mocks__/challenge-nodes';
import mockIntroNodes from '../../../__mocks__/intro-nodes';
import mockCompleted from '../../../__mocks__/completedChallengesMock';

test('<Block /> not expanded snapshot', () => {
  const { container } = render(
    <Block
      blockDashedName='block-a'
      challenges={mockChallengeNodes.filter(node => node.block === 'block-a')}
      completedChallenges={mockCompleted}
      intro={mockIntroNodes[0]}
      isExpanded={false}
      toggleBlock={() => {}}
      toggleMapModal={() => {}}
    />
  );

  expect(container).toMatchSnapshot('block-not-expanded');
});

test('<Block expanded snapshot', () => {
  const { container } = render(
    <Block
      blockDashedName='block-a'
      challenges={mockChallengeNodes.filter(node => node.block === 'block-a')}
      completedChallenges={mockCompleted}
      intro={mockIntroNodes[0]}
      isExpanded={true}
      toggleBlock={() => {}}
      toggleMapModal={() => {}}
    />
  );

  expect(container).toMatchSnapshot('block-expanded');
});

test('<Block />  should handle toggle clicks correctly', async () => {
  const toggleSpy = jest.fn();
  const toggleMapSpy = jest.fn();
  const executeGA = jest.fn();

  const props = {
    blockDashedName: 'block-a',
    challenges: mockChallengeNodes.filter(node => node.block === 'block-a'),
    completedChallenges: mockCompleted,
    intro: mockIntroNodes[0],
    isExpanded: false,
    executeGA: executeGA,
    toggleBlock: toggleSpy,
    toggleMapModal: toggleMapSpy
  };

  const { container, rerender } = render(<Block {...props} />);

  expect(toggleSpy).not.toHaveBeenCalled();
  expect(container.querySelector('.map-title h4')).toHaveTextContent('Block A');
  expect(container.querySelector('ul').children.length).toBe(0);

  fireEvent.click(container.querySelector('.map-title'));

  expect(toggleSpy).toHaveBeenCalledTimes(1);
  expect(toggleSpy).toHaveBeenCalledWith('block-a');

  rerender(<Block {...props} isExpanded={true} />);

  expect(container.querySelector('.map-title h4')).toHaveTextContent('Block A');
  expect(container.querySelector('ul').children.length).toBe(
    props.challenges.length + (props.intro ? 1 : 0)
  );
});
