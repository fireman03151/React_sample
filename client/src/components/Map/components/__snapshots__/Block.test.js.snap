// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Block /> not expanded snapshot: block-not-expanded 1`] = `
<li
  className="block "
>
  <button
    aria-expanded={false}
    className="map-title"
    onClick={[Function]}
  >
    <Caret />
    <h4>
      Block A
    </h4>
    <div
      className="map-title-completed"
    >
      <span>
        <GreenNotCompleted
          style={
            Object {
              "height": "15px",
              "marginRight": "10px",
              "width": "15px",
            }
          }
        />
      </span>
      <span>
        2/4
      </span>
    </div>
  </button>
  <ul />
</li>
`;

exports[`<Block expanded snapshot: block-expanded 1`] = `
<li
  className="block open"
>
  <button
    aria-expanded={true}
    className="map-title"
    onClick={[Function]}
  >
    <Caret />
    <h4>
      Block A
    </h4>
    <div
      className="map-title-completed"
    >
      <span>
        <GreenNotCompleted
          style={
            Object {
              "height": "15px",
              "marginRight": "10px",
              "width": "15px",
            }
          }
        />
      </span>
      <span>
        2/4
      </span>
    </div>
  </button>
  <ul>
    <li
      className="map-challenge-title"
    >
      <span
        className="badge map-badge"
      />
      <Unknown
        onClick={[Function]}
        to="/super-block-one/block-a"
      >
        Introduction to Block A
      </Unknown>
    </li>
    <li
      className="map-challenge-title map-challenge-title-completed"
    >
      <span
        className="badge map-badge"
      >
        <GreenPass
          style={
            Object {
              "height": "15px",
              "marginRight": "10px",
              "width": "15px",
            }
          }
        />
      </span>
      <Unknown
        onClick={[Function]}
        to="/super-block-one/block-a/challenge-one"
      >
        Challenge One
      </Unknown>
    </li>
    <li
      className="map-challenge-title"
    >
      <span
        className="badge map-badge"
      >
        <GreenNotCompleted
          style={
            Object {
              "height": "15px",
              "marginRight": "10px",
              "width": "15px",
            }
          }
        />
      </span>
      <Unknown
        onClick={[Function]}
        to="/super-block-one/block-a/challenge-two"
      >
        Challenge Two
      </Unknown>
    </li>
    <li
      className="map-challenge-title"
    >
      <span
        className="badge map-badge"
      >
        <GreenNotCompleted
          style={
            Object {
              "height": "15px",
              "marginRight": "10px",
              "width": "15px",
            }
          }
        />
      </span>
      <Unknown
        onClick={[Function]}
        to="/super-block-one/block-a/challenge-one"
      >
        Challenge One
      </Unknown>
    </li>
    <li
      className="map-challenge-title map-challenge-title-completed"
    >
      <span
        className="badge map-badge"
      >
        <GreenPass
          style={
            Object {
              "height": "15px",
              "marginRight": "10px",
              "width": "15px",
            }
          }
        />
      </span>
      <Unknown
        onClick={[Function]}
        to="/super-block-one/block-a/challenge-two"
      >
        Challenge Two
      </Unknown>
    </li>
  </ul>
</li>
`;
